### PR TITLE
Add Duplex callback service to WcfTestService (Issue #104)

### PR DESCRIPTION
**Pull request by StephenBonikowsky**

### CODE REVIEW

File: src/System.Private.ServiceModel/src/Extensions/ReflectionExtensions.cs [hunk 1]
```diff
@@ -92,7 +92,7 @@ public static bool IsValueType(this Type type)
         }
         public static InterfaceMapping GetInterfaceMap(this Type type, Type interfaceType)
         {
-            throw ExceptionHelper.PlatformNotSupported();
+            return type.GetTypeInfo().GetRuntimeInterfaceMap(interfaceType);
```
Comment (hongdai) [inline src/System.Private.ServiceModel/src/Extensions/ReflectionExtensions.cs]: How is the product change related to the purpose of this checkin?

Comment (mconnew) [inline src/System.Private.ServiceModel/src/Extensions/ReflectionExtensions.cs]: This is a one line product fix needed to get the test to pass. For pragmatic reasons, it's worth making that one change rather than check in the test, have the test fail, open a new issue, then submit a separate pull request just for this one line.

Comment (iamjasonp) [inline src/System.Private.ServiceModel/src/Extensions/ReflectionExtensions.cs]: Yep - this is a change I made while working on the test+product earlier. 
This change allows us to look for the interface(s) implemented by a given type.

**Comment by iamjasonp** [comment]
Wait, what the heck? I must have managed to send you a patch that wasn't up to my latest commit, as there are changes in there that I no longer have in my branch.

File: src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/Client.ClientBase.Tests.csproj [hunk 2]
```diff
@@ -33,6 +33,7 @@
   </ItemGroup>
   <ItemGroup>
     <Compile Include="ClientBaseTests.cs" />
+    <Compile Include="DuplexClientBaseTests.cs" />
   </ItemGroup>
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.targets))\dir.targets" />
 </Project>
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/Client.ClientBase.Tests.csproj]: Nit: that little red icon says you need to add a newline at the end of the file (even if it was already missing when someone else checked it in.

File: src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/DuplexClientBaseTests.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+using System.ServiceModel.Channels;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+public static class DuplexClientBaseTests
+{
+    static TimeSpan s_WaitTimeout = TimeSpan.FromSeconds(10);
+
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/DuplexClientBaseTests.cs]: Nit: in future we may want this to have a more descriptive name ("MaxTestWaitTime?") and move it into the common scenarios folder.  I suspect we will use it again.  We can wait for the 2nd use before moving it.

File: src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/DuplexClientBaseTests.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        }
+        finally
+        {
+            if (duplexService != null && duplexService.State != CommunicationState.Closed)
+            {
+                duplexService.Abort();
+            }
+        }
+
+        Assert.True(success, "Test timed out.");
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/DuplexClientBaseTests.cs]: I think it would be clearer to move this Assert up right after the WaitOne because "success" only means it didn't time out.  For example, if it does timeout, I'd like to see that assert fire, not the exception that might occur from the Close(). I might consider an Assert to fail if the finally requires an Abort.   That would be useful to know if that path ever executes. Not required -- your call.

File: src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/DuplexClientBaseTests.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+                return _resetEvent;
+            }
+        }
+
+        public void OnPingCallback()
+        {
+            _resetEvent.Set();
+        }
+    }
+}
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/DuplexClientBaseTests.cs]: newline needed

File: src/System.Private.ServiceModel/tests/Scenarios/SelfHostWcfService/Program.cs [hunk 6]
```diff
@@ -12,14 +12,32 @@ private static void Main(string[] args)
...
+
+                DumpEndpoints(wcfservicehost, userNameService, wcfDuplexService);
+
                 Console.WriteLine("Press ENTER to close services...");
                 Console.ReadLine();
             }
         }
+
+         static void DumpEndpoints(params ServiceHost[] hosts)
+         {
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Scenarios/SelfHostWcfService/Program.cs]: Great idea

File: src/System.Private.ServiceModel/tests/Scenarios/SelfHostWcfService/IWcfDuplexService.cs [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+using System.ServiceModel;
+
+namespace WcfService
+{
+    [ServiceContract (CallbackContract = typeof(IWcfDuplexServiceCallback))]
+    public interface IWcfDuplexService
+    {
+        [OperationContract]
+        void Ping();
+    }
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Scenarios/SelfHostWcfService/IWcfDuplexService.cs]: Can we make Ping and OnPingCallback take a parameter, and let Ping echo it back to the callback? The test can ping a GUID and check it gets it back.  Matching up the parameters of the callback operations is an important path to test of the client layer.

File: src/System.Private.ServiceModel/tests/Scenarios/SelfHostWcfService/IWcfDuplexService.cs [hunk 8]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.ServiceModel;
+
+namespace WcfService
+{
+    [ServiceContract (CallbackContract = typeof(IWcfDuplexServiceCallback))]
+    public interface IWcfDuplexService
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Scenarios/SelfHostWcfService/IWcfDuplexService.cs]: FYI, I opened https://github.com/dotnet/wcf/issues/113 to have a shared file for these interfaces that both the tests and selfhost service can share.  We shouldn't have to maintain 2 copies of every interface.  No need to solve that issue with this PR.

File: src/System.Private.ServiceModel/tests/Scenarios/SelfHostWcfService/App.config [hunk 9]
```diff
@@ -151,6 +154,15 @@
                     </baseAddresses>
                 </host>
             </service>
+            <service name="WcfService.WcfDuplexService">
+                <endpoint address="tcp-nosecurity-callback" binding="netTcpBinding" bindingConfiguration="nettcpBindingCallback"
+                          name="nettcp-defaultBinding" contract="WcfService.IWcfDuplexService" />
+                <host>
+                    <baseAddresses>
+                        <add baseAddress="net.tcp://*:810/WindowsCommunicationFoundation" />
```
Comment (iamjasonp) [inline src/System.Private.ServiceModel/tests/Scenarios/SelfHostWcfService/App.config]: In the latest patch, there should be no need to set up a new baseAddress. I had merged it back into the same endpoint for simplicity (i.e., `*:809`).

File: src/System.Private.ServiceModel/tests/Scenarios/SelfHostWcfService/IWcfDuplexService.cs (hunk 7 above)
Comment (iamjasonp) [inline src/System.Private.ServiceModel/tests/Scenarios/SelfHostWcfService/IWcfDuplexService.cs]: :+1:

File: src/System.Private.ServiceModel/tests/Scenarios/SelfHostWcfService/Program.cs (hunk 6 above)
Comment (iamjasonp) [inline src/System.Private.ServiceModel/tests/Scenarios/SelfHostWcfService/Program.cs]: :+1: on `DumpEndpoints`. 

Not important: Is there a way that `DumpEndpoints` can just enumerate through all ServiceHosts present on the system (or app domain) rather than have to pass this information in?

Comment (SajayAntony) [inline src/System.Private.ServiceModel/tests/Scenarios/SelfHostWcfService/Program.cs]: @iamjasonp I don't think its possible without walking the heap or attaching something like an on open handler. I just did this to keep it simple and get an output of all the endpoints when the exe starts up.

**Comment by mconnew** [comment]
This isn't the correct thing to review. A git rebase master dropped one of Stephen's commits. I've recovered it for him but it will be tomorrow before he can update this PR.

File: src/System.Private.ServiceModel/src/Extensions/ReflectionExtensions.cs (hunk 1 above)
Comment (iamjasonp) [inline src/System.Private.ServiceModel/src/Extensions/ReflectionExtensions.cs]: To clarify: the test isn't passing for me right now, I need that one line change to get past the `PlatformNotSupportedException`.

**Comment by SajayAntony** [comment]
:shipit:

**Comment by roncain** [comment]
LGTM
